LandSandBoat/server
# Lua, SQL: Implement Zipacna pathing and interaction with Cermet Doors

**_I affirm:_**

## What does this pull request do?


Currently, Zipacna only spawns in one place and paths through the doors regardless if they are open or closed.
This PR addresses the following: (I made no modifications to Zipacna's abilities or stats)
- adds the 4 spawn points located as shown here.
![393557417-d9385703-103a-4d8a-95d6-3f5bc5cd3a24](https://github.com/user-attachments/assets/3f44d802-d75a-4519-b4f9-2ecbb134187b)
![393557420-833a10fd-b3bd-4b09-8b42-36c9a87db275](https://github.com/user-attachments/assets/6696dab4-5e9b-4378-a52a-ddbe29f8ab0e)
![393557428-58f35a18-16f1-4345-9ae7-99b6331d1983](https://github.com/user-attachments/assets/3640562f-110d-4adf-a8c2-0a30410d2b86)

    https://www.youtube.com/watch?v=nWjGMWHsg9k

The doors are labelled as such (from west side to east side)
Basement Loop - |1st Blue Door | Middle Blue Door | Middle Yellow Door | 1st Yellow Door| - Basement loop

- There are 4 doors controlled by a Yellow or Blue Monolith. When one is activated the other "side doors" are closed.
** Note - It is possible that Zipacna can get stuck between the middle yellow and first yellow door as well as the middle blue door and first blue door. If someone switches while he is pathing between **

Zipacna's behavior is to path to a door. If the door is closed, it will turn around and go to the next door in the sequence.
Example: Middle Yellow door was closed, Zipacana will turn around and go to the middle blue door. If the middle blue door is open, Zipacna will proceed to the 1st Blue door, make a loop in the basement and then return to the first blue door. If however, the yellow switch is triggered and the middle blue door is now closed. once Zipacna arrives at the door again, he will turn around and go to the middle yellow door.

- Removed Hard-Coded Position Checks (x/y/z)
- Updated to use `GetNPCByID()` and `getPos()` to get positions on doors
- Updated to use `checkDistance()` when needed
- Updated to use `xi.path.flag.REVERSE` where applicable to remove some of the extra path nodes
- Converted that path nodes into a table instead of several local variables
- Created a file-based scope variable `currentDirection` to track the direction of Zipacna (vs using a Local Var on the mob)
- Extracted the 4x replicated door logic into a helper function `setZipPath()` to handle pathing logic
- Fixed all comment formatting errors
- Removed un-needed local vars on Zipacna
- Removed excess prints that were left in the lua file

## Steps to test these changes


1 - `!spawnmob 17502570`
2 - `!gotoid 17502570`
3 - Follow Zipacna through the zones and up to blue/yellow doors
Note that he now paths to the doors, and if closed, turns around and paths the opposite direction.
4 - Trigger the doors to open and close so that he can path to the different segments
Note that even if he is 'trapped' between two doors, he will just path back and forth between the two doors until he finds an open door to continue.
